### ISSUE TITLE
Error in while (i <= maxiter && criterion_val > tol)

### ISSUE THREAD

**Issue by RaphaelHartmann** [quentingronau/bridgesampling#13]
Dear all,

I hope this is the right place to post the following issue:

I tried to run the `bridge_sampler()` function as follows:
```
bridge_sampler(samples = bridge[,-npar],
               log_posterior = "unnorm_logpost", data = data_list, lb = lb, ub = ub,
               repetitions = 1, param_types = rep("real", (npar-1)),
               method = "normal", cores = 1, use_neff = TRUE, packages = NULL,
               varlist = NULL, envir = .GlobalEnv, rcppFile = NULL,
               maxiter = 200, silent = FALSE, verbose = FALSE)
```
where `bridge` is a self constructed `mcmc.list` using `as.mcmc.list()`. With `-npar` I just exclude the last column in the mcmc list, since this column contains the log-likelihood. The `log_posterior` function is a little special, since I am using a .dll file and the `.Call()` function to calculate the unnormalized log-posterior:

```
unnorm_logpost <- function(samples.row, data) {
  REAL <- as.numeric(samples.row)
  REAL2 <- as.numeric(data$ConstProb)
  CHAR <- c(data$dat_mod)
  INTEGER <- as.integer(data$numbers)
  INTEGER2 <- as.integer(data$CatToResp)
  BOOL1 <- as.logical(data$lambda_minus)
  BOOL2 <- as.logical(data$lambda_plus)

  out_bridge <- .Call("bridge", REAL, REAL2, CHAR, INTEGER, INTEGER2, BOOL1, BOOL2)
  return(out_bridge$unnorm_log_post)
}
```
When I run the above `bridge_sampler()` function I get the output:

```
...
Iteration: 196
Iteration: 197
Iteration: 198
Iteration: 199
Iteration: 200
Iteration: 1
```

and the following error message:

```
Error in while (i <= maxiter && criterion_val > tol) { : 
  missing value where TRUE/FALSE needed
```

and in addition the usual warning when using too few iterations (in my case less than about 240 iterations):

```
logml could not be estimated within maxiter, rerunning with adjusted starting value. 
Estimate might be more variable than usual. 
```

This seems to be a little random, but if I use only about 100 or less as `maxiter` I get no error message... only the warning. I think this is only by chance, but still it looks like it should work.

I also tried to capture NaNs, NAs, and infinity values for the output in my `log_posterior` function with `is.nan()`, `is.na()`, `is.infinite()`, and `stop()` after the `.Call()`, but there seem to be no invalid values produced by my function. The code for that is not shown above.

I could reproduce the error behaviour by using `packrat` and installing `bridgesampling` "anew". 

Example 1 and 2 from the package description run fine. So I think it is not likely to be due to my R installation.

Any help / suggestions would be appreciated.

Best wishes
Raphael

**Comment by quentingronau** [comment]
Hi Raphael,

Thanks for reporting this. Could you please provide us with a [reproducible example](https://stackoverflow.com/questions/5963269/how-to-make-a-great-r-reproducible-example) so that we could investigate this?

Cheers,
Quentin

**Comment by RaphaelHartmann** [comment]
Dear Quentin Gronau

Here is a link to a zip-file containing files you need to produce the error:
https://osf.io/e8pg2/

It is password-protected since I use a real data set. The password is:
thx4yourHELP

To run the example just open the file bridge.R and execute.

Some infos: The BRIDGE.dll file of course works only on Windows. I don't have a .so file yet. It should be usable with R version 3.4.4 and above.
If there are any questions about the code or other infos I forgot to tell feel free to cantact me here.

Thank you for helping me and sorry for not providing the example before.

Best wishes
Raphael

Sent with [ProtonMail](https://protonmail.com) Secure Email.

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐

**Comment by RaphaelHartmann** [comment]
Hi Quentin

I hope you saw my second e-mail to your e-mail adress (quentin.f.gronau@gmail.com) with a link to the password protected zip-file.

Best wishes
Raphael

**Comment by quentingronau** [comment]
Hi Raphael,

Yes, thanks. Since I am unfortunately unable to open the files on a Windows computer, I forwarded your mail to @singmann who is also an author of the bridgesampling package.

Cheers,
Quentin

**Comment by singmann** [comment]
I will try to look into it asap, but could take until the weekend.

**Comment by RaphaelHartmann** [comment]
But the password works, right?

Thank you :) 
No stress.

Best wishes

**Comment by RaphaelHartmann** [comment]
Any progress so far?

**Comment by singmann** [comment]
Even with your `.dll` I am not getting anywhere:

```
> dyn.load("tmp/BRIDGE.dll")
Error in inDL(x, as.logical(local), as.logical(now), ...) : 
  unable to load shared object 'C:/Users/henrik/Documents/GitHub/bridgesampling/tmp/BRIDGE.dll':
  LoadLibrary failure:  The specified module could not be found.
```

Please note that the path is of course correct.

My system:

```
> sessionInfo()
R version 3.5.1 (2018-07-02)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

Matrix products: default

locale:
[1] LC_COLLATE=English_United Kingdom.1252  LC_CTYPE=English_United Kingdom.1252   
[3] LC_MONETARY=English_United Kingdom.1252 LC_NUMERIC=C                           
[5] LC_TIME=English_United Kingdom.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_3.5.1 tools_3.5.1    yaml_2.2.0  
```

In any case, most problems with bridgesampling result from using too few posterior samples. Can you confirm that you have a lot of posterior samples? And with a lot I mean at least an order of magnitude more than required for estimation. From your file it seems like you only have 5000 * 4 posterior samples. That probably is not enough. Especially for a model with 655 parameters. Maybe try 20,000 times 4 or even 50,000 times 4. If the problem then still persists, it might be worth investing the time to getting it run on my machine.

You could also try to send us the

**Comment by RaphaelHartmann** [comment]
I will try it out with more samples and report you whether there are still problems.

I attached a zip with a dll. When you load that, you should be able to `.Call` "bridge" like before. This dll is build under R 3.4.4:
[rtmpt.zip](https://github.com/quentingronau/bridgesampling/files/2679993/rtmpt.zip)
```
> sessionInfo()
R version 3.4.4 (2018-03-15)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

Matrix products: default

locale:
[1] LC_COLLATE=German_Switzerland.1252  LC_CTYPE=German_Switzerland.1252    LC_MONETARY=German_Switzerland.1252
[4] LC_NUMERIC=C                        LC_TIME=German_Switzerland.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_3.4.4 tools_3.4.4 
```
I hope this helps. If you can load it, you should be able to run the script I sent you.

Best wishes
Raphael

**Comment by RaphaelHartmann** [comment]
Here the new zip file:
[rtmpt.zip](https://github.com/quentingronau/bridgesampling/files/2689786/rtmpt.zip)
If I use 30000x4 samples it stops at iteration `i < 200` with `maxiter = 1000 `and I get the following error:
```
Error in while (i <= maxiter && criterion_val > tol) { : 
  missing value where TRUE/FALSE needed
```

If I use 40000x4 samples I get the following error without any iteration done:
```
Error in eigen(if (doDykstra) R else Y, symmetric = TRUE) : 
  infinite or missing values in 'x'
In addition: There were 12 warnings (use warnings() to see them)
```
I cannot access the warnings since it is stuck there.

This is my `log_posterior()` function:
```
unnorm_logpost <- function(samples.row, data) {
  REAL <- as.numeric(samples.row)
  REAL2 <- as.numeric(data$ConstProb)
  CHAR <- c(data$dat_mod)
  INTEGER <- as.integer(data$numbers)
  INTEGER2 <- as.integer(data$CatToResp)
  BOOL1 <- as.logical(data$lambda_minus)
  BOOL2 <- as.logical(data$lambda_plus)
  out_bridge <- .Call("bridge", REAL, REAL2, CHAR, INTEGER, INTEGER2, BOOL1, BOOL2)
  if (is.na(out_bridge$unnorm_log_post)) stop("NA produced!\n")
  if (is.infinite(out_bridge$unnorm_log_post)) stop("Inf produced!\n")
  return(out_bridge$unnorm_log_post)
}
```

Every log-likelihood and unnormalized log posterior is printed out from within C++.

Best wishes
Raphael

**Comment by hstojic** [comment]
Hello,

has there been any clue to these errors? I encounter the same errors with a model 
```R
Error in while (i <= maxiter && criterion_val > tol) { :
  missing value where TRUE/FALSE needed
```

i didn't try running a much larger niter at the moment as the model fitting takes a long time as it is. What is the source of the error, i.e. what data should I look at?

Otherwise, looking at the diagnostics, model seems OK.

**Comment by singmann** [comment]
Hey Hrvoje,

Unfortunately, we did not yet receive a reproducible example to investigate what is going wrong. If you have one, we might be able to figure this out. 

But again, your best bet is probably to increase the number of iterations from the MCMC sampler. I know it might be super time consuming, but the general problem is that obtaining the marginal likelihood i just a very difficult numerical problem. So you need a lot of data to do this reliably. Especially in high-dimensional problems. And if you see the number of iterations of the bridge sampler going up and then this happening, it might suggest that the amount of information you provide is still not enough.

Just for reference, I have a model for 3232 data points with roughly 2000 parameters (which includes of course hierarchical structures and covariance parameters). To estimate the marginal likelihood reliably, I run 12 parallel chains, each with 80000 posterior samples with thinning of 2. So in the end I have almost 500000 posterior samples which I pass to bridge_sampler. And these are samples from Stan so the ratio of effective samples to posterior samples is pretty high.

I know this many are needed, because I tried less and then it either failed (potentially with similar error messages, I do not remember), or more importantly, the marginal likelihoods were unreliable across repeated estimation attempts. That means, I obtained independent sets of posterior samples and estimated the marginal likelihood for each set (actually twice for each set). And only with this many samples did the marginal likelihood from independent sets of posterior samples were within acceptable ranges, the range of different log-ml estimates for each model was less than 1 (for my set of candidate models this was enough to reliably distinguish between models).

**Comment by singmann** [comment]
We have finally addressed and hopefully solved this problem in 111841faeb40ee9664fac71a5fa547b4ea906dcc. This error message should not appear again with `bridgesampling` version `0.7-1` or larger. If not, please post another reproducible example.

**Comment by singmann** [comment]
The new version is meanwhile on CRAN.
